pydantic/dataclasses.py

```diff
@@ -211,3 +211,8 @@ def create_dataclass(cls: type[Any]) -> type[PydanticDataclass]:
 
 
 __getattr__ = getattr_migration(__name__)
+
+if (3, 8) <= sys.version_info < (3, 11):
+    # Monkeypatch dataclasses.InitVar so that typing doesn't error if it occurs as a type when evaluating type hints
+    # This is not necessary in 3.11, and we don't support InitVar in python < 3.8.
+    dataclasses.InitVar.__call__ = lambda: None  # type: ignore
```

Comment:
What happens if this gets used before pydantic gets imported?

Reply:
It would error if you tried to use `typing.get_type_hints` first, but currently I don't think there is a way to get to the point of building a dataclass schema without importing something from `pydantic.dataclasses`, at least the following works:
```import dataclasses

from pydantic import BaseModel

initvars = []
@dataclasses.dataclass
class A:
    x: 'dataclasses.InitVar[int]'
    def __post_init__(self, x):
        initvars.append(x)

class B(BaseModel):
    a: A

B.model_validate({'a': {'x': 2}})

assert initvars == [2]

B.model_validate({'a': {'x': 'abc'}})  # errors
```

Do you think we need to be more careful about ensuring the monkeypatch happens no matter what import path is followed?

Reply:
`InitVar.__call__` doesn't actually get called — the `typing.get_type_hints` function errors if it doesn't think the annotations are actually types. It doesn't think `InitVar[int]` is a type pre-3.11 I guess, but one thing you can do to get it to make it think something is a type is to make that thing callable (it has a boolean check `if callable(maybe_type):` in which case it lets you use that annotation and not error). That's why this monkeypatch works — as long as `get_type_hints` doesn't get called on the type until after the monkeypatch is applied (which I think is the case because that doesn't happen until you are building the schema).